### PR TITLE
Add functionality for per-layer clipping

### PR DESCRIPTION
**Pull request by notimeforcaution**
Per-layer, translates to per-leaf in the PyTree based implementation. Currently, only a singular value of max L2 Norm per leaf is supported. Upcoming change will allow the users to provide max L2-norm per leaf as a PyTree, where the L2 Norm PyTree will be a "prefix tree" of the Gradient PyTree.

### CODE REVIEW

**Comment by notimeforcaution** [comment]
Results of lint-based checks:

```
✦ ~/Documents/jax_privacy add-per-layer-clipping*
venv ❯ pytype tests/clipping_test.py
Computing dependencies
Analyzing 1 sources with 0 local dependencies
ninja: Entering directory `.pytype'
[1/1] check clipping_test
Leaving directory '.pytype'
Success: no errors found

✦ ~/Documents/jax_privacy add-per-layer-clipping*
venv ❯ flake8 jax_privacy/**.py      

✦ ~/Documents/jax_privacy add-per-layer-clipping*
venv ❯ pylint jax_privacy/**.py

✦ ~/Documents/jax_privacy add-per-layer-clipping* 6s
venv ❯ pytype jax_privacy/**.py
Computing dependencies
Analyzing 12 sources with 10 local dependencies
ninja: Entering directory `.pytype'
[16/16] check jax_privacy.__init__
Leaving directory '.pytype'
Success: no errors found

✦ ~/Documents/jax_privacy add-per-layer-clipping* 22s
venv ❯ flake8 tests/clipping_test.py

✦ ~/Documents/jax_privacy add-per-layer-clipping*
venv ❯ pylint tests/clipping_test.py
************* Module clipping_test
tests/clipping_test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
tests/clipping_test.py:205:10: W0212: Access to a protected member _adjusted_clip_norm of a client class (protected-access)
tests/clipping_test.py:213:6: W0212: Access to a protected member _adjusted_clip_norm of a client class (protected-access)

✦ ~/Documents/jax_privacy add-per-layer-clipping*
venv ❯ pytype tests/clipping_test.py
Computing dependencies
Analyzing 1 sources with 0 local dependencies
ninja: Entering directory `.pytype'
[1/1] check clipping_test
Leaving directory '.pytype'
Success: no errors found
```

There are some lint errors in unrelated lines in tests/ directory. In order to keep changes isolated I am avoiding to change these lines in this PR. I will send some follow-up changes to fix these lint errors.

File: jax_privacy/clipping.py [hunk 1]
```diff
@@ -584,6 +621,7 @@ def clipped_grad(
     has_aux: bool = False,
     *,
     l2_clip_norm: float,
+    per_layer: bool = False,
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Per your proposal document, I think it's actually better to not add new arguments here, but just allow l2_clip_norm to be a pytree of floats instead.  I thought this is what you initially proposed

**Review by ryan112358 — CHANGES_REQUESTED** [review]

File: jax_privacy/clipping.py (hunk 1 above)
Comment (notimeforcaution) [reply jax_privacy/clipping.py]: My thinking earlier was, to have 'per layer' arg, and let l2_clip_norm both. Then, let the latter either be a Real Number (in case we wish to apply same clip norm value on each leaf) or a PyTree (in case we wish to apply separate clip norm values on different leaves). Now, upon looking closely, this does feel overengineered.

Your suggestion (IIUC) is:
- Removing per_layer
- Allowing l2_clip_norm to accept a PyTree or a RealNumber. In case its PyTree we do a per layer clipping, else we do a Global Clipping.

That actually looks very good to me, and helps us remove a potential redundancy in args. I will make these changes, and update. If you think my understanding of your suggestion is incorrect, feel free to suggest that as well.

File: jax_privacy/clipping.py [hunk 2]
```diff
@@ -139,16 +145,60 @@ def clip_pytree(
...
-  clip_norm = jnp.maximum(clip_norm, 0.0)
-  l2_norm = optax.tree.norm(pytree)
-  scale = jnp.minimum(1.0, clip_norm / l2_norm)
-  if rescale_to_unit_norm:
-    scale = jax.lax.select(clip_norm > 0, scale / clip_norm, 1 / l2_norm)
-  # If l2_norm is 0 or nan, set scale to 0.0.
-  scale = jnp.nan_to_num(scale, nan=0.0, posinf=0.0)
-  clipped = jax.tree.map(lambda x: jnp.astype(scale, x.dtype) * x, pytree)
+
+  def _calculate_scale(l2_norm, c_norm):
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: This is significantly more complicated than the original.  Can you try to simplify it?  I think you could set it up as a 2-layer recursive call.  If the l2_clip_norm is not a numeric type / jax.Tracer -> call jax.tree.map ove rthe leaves of clip norm.  Otherwise, keep the current implementation.  This would be a much lighter weight change I'd be more comfortable accepting

File: jax_privacy/clipping.py [hunk 3]
```diff
@@ -522,16 +579,6 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
 
     clipped_values, aux, norms = microbatched_vmap_fun(*args, **kwargs)
     if keep_batch_dim:
-      # If keep_batch_dim is True, we artificially added a dimension of size 1
```
Comment (ryan112358) [inline jax_privacy/clipping.py:524]: Avoid changing unrelated code

File: jax_privacy/clipping.py [hunk 4]
```diff
@@ -465,6 +516,14 @@ def clipped_fun(
           'normalize_by is not compatible with grid_scale. Normalization '
           'should be applied after noise addition.'
       )
+    is_global_clip = jax.tree_util.treedef_is_leaf(
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: jax.tree_util is an old pattern I think, it's more common to use jax.tree.is_leaf, although I'm not familiar with this symbol.  Can you just check if l2_clip_norm is not a numeric type?

File: jax_privacy/clipping.py [hunk 5]
```diff
@@ -465,6 +516,14 @@ def clipped_fun(
...
+        jax.tree_util.tree_structure(l2_clip_norm)
+    )
+    if not is_global_clip:
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I recently added _validate.py, maybe this check can go there as "is_numeric" or something

File: jax_privacy/clipping.py [hunk 6]
```diff
@@ -741,9 +811,17 @@ def grad_fn(*args, **kwargs):
     value_and_aux, grad = value_and_grad_fn(*args, **kwargs)
     result = pre_clipping_transform(grad)
     if has_aux or return_values or return_grad_norms:
+      if return_grad_norms:
+        if isinstance(l2_clip_norm, numbers.Real):
+          grad_norms_val = optax.tree.norm(grad)
+        else:
+          grad_norms_val = jax.tree.map(jnp.linalg.norm, grad)
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: This gives you the per-leaf norms, but if l2_clip_norm is a prefix of the parameter tree, then you might want the grad norms to match that structure.

File: jax_privacy/clipping.py [hunk 7]
```diff
@@ -741,9 +811,17 @@ def grad_fn(*args, **kwargs):
     value_and_aux, grad = value_and_grad_fn(*args, **kwargs)
     result = pre_clipping_transform(grad)
     if has_aux or return_values or return_grad_norms:
+      if return_grad_norms:
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I think rather than all this nesting, just do this

grad_norms = jax.tree.map(lambda _, g: optax.tree.norm(g), l2_clip_norm, grad)

Down below you can still have the check if "return_grad_norms" is True.  If it is False, then JIT will compile-away this dead code, so you don't need the extra level of nested conditional.

File: jax_privacy/clipping.py [hunk 8]
```diff
@@ -666,7 +734,9 @@ def clipped_grad(
       Exercise caution when using this as no DP sensitivity guarantees are
       provided for the auxiliary data.
     l2_clip_norm: The maximum L2 norm for each per-example gradient. Gradients
-      with a norm larger than this value will be scaled down.
+      with a norm larger than this value will be scaled down. Can be a single
```
Comment (ryan112358) [inline jax_privacy/clipping.py:722]: The conventions of return_grad_norm also changed.  Be sure to add a sentence to the doc about that.

File: tests/clipping_test.py [hunk 9]
```diff
@@ -54,7 +54,10 @@ def test_clip_pytree_output_dtype_matches_input(self, pytree, **kwargs):
     pytree = optax.tree.random_like(jax.random.key(0), pytree)
     clipped, _ = clipping.clip_pytree(pytree, **kwargs)
 
-    jitted = jax.jit(clipping.clip_pytree, static_argnums=(2, 3))
```
Comment (ryan112358) [inline tests/clipping_test.py:57]: This seems unrelated

File: jax_privacy/clipping.py [hunk 10]
```diff
@@ -741,9 +811,17 @@ def grad_fn(*args, **kwargs):
...
+      else:
+        grad_norms_val = None
+
       aux = AuxiliaryOutput(
           values=value_and_aux[0] if return_values else None,
-          grad_norms=optax.tree.norm(grad) if return_grad_norms else None,
+          grad_norms=grad_norms_val,
           aux=value_and_aux[1] if has_aux else None,
       )
       return result, aux
```
Comment (ryan112358) [inline jax_privacy/clipping.py:815]: You will also have to updat ethe "l2_norm_bound" of the BoundedSensitivityCallable.  Now, it should be the l2 norm of the leaves of l2_clip_norm, I guess l2_norm_bound=optax.tree.norm(l2_clip_norm) is the right value

File: tests/clipping_test.py [hunk 11]
```diff
@@ -158,6 +161,98 @@ def test_clip_pytree_with_inf(self):
...
+        'extra_key': invalid_clip_extra_key,
+        'type_mismatch': invalid_clip_type_mismatch,
+    }
+
+    for name, invalid_clip in mismatch_cases.items():
+      with self.subTest(name=name):
+        with self.assertRaises(ValueError):
+          clipping.clip_pytree(pytree, invalid_clip)
+
+  def test_clip_pytree_with_prefix_tree_clip_norm(self):
```
Comment (ryan112358) [inline tests/clipping_test.py]: See if you can merge some of these tests into a smaller number of parameterized tests.  You could sweep l2_clip_norm over a few values covering scalar, pytree-matching-leaves, and pytree-prefix.  Maybe it can even be added to an existing test, not sure.

File: tests/clipping_test.py [hunk 12]
```diff
@@ -434,6 +539,84 @@ def fun(batched_data):
     self.assertEqual(aux_3d.shape, (15, 4, 9))
     self.assertEqual(aux_4d.shape, (15, 4, 9, 10))
 
+  def test_return_norms_per_layer_structure(self):
```
Comment (ryan112358) [inline tests/clipping_test.py]: If you do the parameterized test above, just fold this check in there rather than creating a completely new test.  chex.assert_tree_structure_equal(l2_clip_norm, grad_norms) should suffice (or something like it).

**Review by ryan112358 — CHANGES_REQUESTED** [review]

File: jax_privacy/clipping.py (hunk 2 above)
Comment (notimeforcaution) [reply jax_privacy/clipping.py]: Yes, completely agreed!

---

Side Note: This PR is a WIP, after changes to the initial proposed API design. In this version, it is purely functional (and not production-ready) code. I wanted to clean changes, and optimize wherever feasible before sending for review. However, I saw that it was marked as Open (and not Draft); and I am sorry for the same.

I have marked this as Draft now, and am working on addressing comments, and cleaning the logic. I will send it for review just as that is done.

File: tests/clipping_test.py (hunk 9 above)
Comment (notimeforcaution) [reply tests/clipping_test.py:57]: Yes, to be deleted. It was originally there because in earlier iteration I had added "per_layer" as a separate argument, and this order was stale. Considering that we are not adding new arguments now, this can be deleted.

File: jax_privacy/clipping.py (hunk 4 above)
Comment (notimeforcaution) [reply jax_privacy/clipping.py]: Done.

Used `jnp.isscalar(clip_norm)`, which returns true when clip_norm is a tracer or scalar; works best here!

File: jax_privacy/clipping.py (hunk 5 above)
Comment (notimeforcaution) [reply jax_privacy/clipping.py]: jnp.isscalar does the exact same thing (https://docs.jax.dev/en/latest/_autosummary/jax.numpy.isscalar.html#jax-numpy-isscalar), so going ahead with that. We probably won't require a custom check.

File: jax_privacy/clipping.py [hunk 13]
```diff
@@ -132,6 +145,24 @@ def clip_pytree(
...
+          leaf_clip_norm,
+          rescale_to_unit_norm=rescale_to_unit_norm,
+          nan_safe=nan_safe,
+          return_zero=return_zero,
+      )
+
+    results = jax.tree.map(recursive_clip, clip_norm, pytree)
+    clipped = jax.tree.map(lambda _, r: r[0], clip_norm, results)
+    leaf_norms = jax.tree.map(lambda _, r: r[1], clip_norm, results)
+    return clipped, optax.tree.norm(leaf_norms)
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I actually thing the contract of this function should be to return the clipped pytree along with the leaf norms

File: jax_privacy/clipping.py [hunk 14]
```diff
@@ -132,6 +145,24 @@ def clip_pytree(
     A tuple `(clipped_pytree, original_l2_norm)`, where `clipped_pytree` is the
     processed PyTree and `original_l2_norm` is the L2 norm of the input PyTree.
   """
+  # -- Per-layer Norm --
+  if not jnp.isscalar(clip_norm):
+
+    def recursive_clip(leaf_clip_norm, leaf_tree):
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Consider using functools.partial here

File: jax_privacy/clipping.py [hunk 15]
```diff
@@ -547,8 +584,17 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
 
   if grid_scale is not None:
     norm_bound = float(grid_scale)
+  elif rescale_to_unit_norm:
+    # If every leaf is rescaled to unit norm, the worst-case global norm is
+    # the square root of the total number of leaves in the PyTree.
+    num_leaves = len(jax.tree_util.tree_leaves(l2_clip_norm))
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: prefer jax.tree.leaves over jax.tree_util.tree_leaves

File: jax_privacy/clipping.py [hunk 16]
```diff
@@ -547,8 +584,17 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
...
+    norm_bound = math.sqrt(num_leaves) / normalize_by
   else:
-    norm_bound = (1.0 if rescale_to_unit_norm else l2_clip_norm) / normalize_by
+    norm_bound = optax.tree.norm(l2_clip_norm) / normalize_by
+    # Cast to a float only when norm_bound is not a JAX Tracer.
+    if not isinstance(norm_bound, jax.core.Tracer):
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Ah this is interesting.  I guess optax.tree.norm will convert the output to a jax array even if the input is define w.r.t. primitives.  You'll probably have to write your own _primitive_tree_norm helper function using a combination of jax.tree.flatten and np.linalg.norm (numpy not jax).

Actually np.linalg.norm will assume the input is always primitive.  Technically the l2_clip_norm could be either primitive or jax scalars, so you need to avoid np.linalg.norm and manually do sum(leaf**2 for leaf in jax.tree.flatten(l2_clip_norm)).  Since it's a 1-liner, you might not need a helper function.

File: jax_privacy/clipping.py [hunk 17]
```diff
@@ -607,6 +653,15 @@ def clipped_grad(
   gradient to have a norm of at most `l2_clip_norm`, and finally sums these
   clipped gradients.
 
+  The clipping behavior is determined by the type of `l2_clip_norm`:
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I'd remove this documentation from here, and just expand the documentatoin of the l2_clip_norm Arg below

File: tests/clipping_test.py [hunk 18]
```diff
@@ -434,6 +440,118 @@ def fun(batched_data):
     self.assertEqual(aux_3d.shape, (15, 4, 9))
     self.assertEqual(aux_4d.shape, (15, 4, 9, 10))
 
+    @parameterized.named_parameters(
```
Comment (ryan112358) [inline tests/clipping_test.py]: I believe there are some property-based tests in this file that you can expand to increase test coverage as well . It would probably be a bit cleaner than this huge parameterized test

**Review by ryan112358 — CHANGES_REQUESTED** [review]
Overall it's looking quite clean (except the test), nice work!

File: jax_privacy/clipping.py (hunk 16 above)
Comment (ryan112358) [reply jax_privacy/clipping.py]: sum(leaf**2 for leaf in jax.tree.flatten(l2_clip_norm)[0])**0.5

File: jax_privacy/clipping.py [hunk 19]
```diff
@@ -132,6 +145,24 @@ def clip_pytree(
...
+      return clip_pytree(
+          leaf_tree,
+          leaf_clip_norm,
+          rescale_to_unit_norm=rescale_to_unit_norm,
+          nan_safe=nan_safe,
+          return_zero=return_zero,
+      )
+
+    results = jax.tree.map(recursive_clip, clip_norm, pytree)
+    clipped = jax.tree.map(lambda _, r: r[0], clip_norm, results)
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: This could be written as jax.tree.transpose(jax.tree.structure(pytree), None, results)

File: jax_privacy/clipping.py [hunk 20]
```diff
@@ -131,7 +144,28 @@ def clip_pytree(
...
+        nan_safe=nan_safe,
+        return_zero=return_zero,
+    )
+    results = jax.tree.map(
+        lambda clip_norm, pytree: recursive_clip(pytree, clip_norm),
+        clip_norm,
+        pytree,
+    )
+    is_tuple = lambda x: isinstance(x, tuple)
+    clipped = jax.tree.map(lambda x: x[0], results, is_leaf=is_tuple)
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: The "is_tuple" check is brittle, because the pytree itself could contain a real tuple that is not a leaf.  Would jax.tree.transpose work here?

File: jax_privacy/clipping.py [hunk 21]
```diff
@@ -142,11 +176,11 @@ def clip_pytree(
   l2_norm = optax.tree.norm(pytree)
   scale = jnp.minimum(1.0, clip_norm / l2_norm)
   if rescale_to_unit_norm:
-    scale = jax.lax.select(clip_norm > 0, scale / clip_norm, 1 / l2_norm)
+    scale = jnp.where(clip_norm > 0, scale / clip_norm, 1 / l2_norm)
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: This seems unrelated to the current PR, why change?

File: tests/clipping_test.py [hunk 22]
```diff
@@ -100,40 +129,89 @@ def test_clip_pytree_clip_norm_zero_rescale(self):
...
       *cartesian_product(
-          clip_norm=[0.0, 1.0, 2.0, jnp.inf],
+          clip_norm=[
+              0.0,
+              1.0,
+              2.0,
+              jnp.inf,
+              {
+                  'a': np.array(jnp.inf),
+                  'b': np.array(jnp.nan),
```
Comment (ryan112358) [inline tests/clipping_test.py]: Nan in clipping norm is not an edge case I've thought about, what's the intended behavior there?

File: tests/clipping_test.py [hunk 23]
```diff
@@ -100,40 +129,89 @@ def test_clip_pytree_clip_norm_zero_rescale(self):
...
+                  'c': np.array(jnp.inf),
+              },
+          ],
           return_zero=[False, True],
           rescale_to_unit_norm=[False, True],
       )
   )
   def test_nan_safety(self, **kwargs):
-    pytree = jnp.array([3.0, 0.0, 1.0, jnp.nan, 2.0])
```
Comment (ryan112358) [inline tests/clipping_test.py:131]: Just make this a two-element dict pytree by default for all tests and avoid the if / else

File: tests/clipping_test.py [hunk 24]
```diff
@@ -100,40 +129,89 @@ def test_clip_pytree_clip_norm_zero_rescale(self):
     chex.assert_trees_all_close(clipped, expected)
 
   @parameterized.parameters(
-      *cartesian_product(pytree=PYTREE_STRUCTS, clip_norm=[0.0, 1.0, jnp.inf])
+      *cartesian_product(
```
Comment (ryan112358) [inline tests/clipping_test.py]: Try to keep the existing parameterized annotation on one line, here and elsewhere , e.g., clip_norm=[0.0, 1.0, jnp.inf, {'a': 1, 'b': 2}]

File: tests/clipping_test.py [hunk 25]
```diff
@@ -61,22 +61,51 @@ def test_clip_pytree_output_dtype_matches_input(self, pytree, **kwargs):
...
+              },
+          ],
           rescale_to_unit_norm=[True, False],
           nan_safe=[True, False],
           return_zero=[True, False],
       )
   )
   def test_clip_pytree_has_bounded_norm(self, pytree, **kwargs):
+    # Expect a ValueError when pytree is a leaf, and clip_norm is a PyTree.
+    if isinstance(kwargs['clip_norm'], dict) and not isinstance(pytree, dict):
```
Comment (ryan112358) [inline tests/clipping_test.py]: These if / else are too complicated, please simplify

File: tests/clipping_test.py [hunk 26]
```diff
@@ -61,22 +61,51 @@ def test_clip_pytree_output_dtype_matches_input(self, pytree, **kwargs):
   @parameterized.parameters(
       *cartesian_product(
           pytree=PYTREE_STRUCTS,
-          clip_norm=[0.0, 1.0, 2.0, jnp.inf],
+          clip_norm=[
```
Comment (ryan112358) [inline tests/clipping_test.py]: Please keep clip_norm= to one line

File: jax_privacy/clipping.py (hunk 20 above)
Comment (ryan112358) [reply jax_privacy/clipping.py]: Perhaps the safest thing to do would be to define a _ClippedTreeAndNorm namedtuple.  so you can check against that specifically.  But given that this is a tricky point, I would say you should create a new parameterized tests of clip pytree that hits it with different combinations of parameters / l2_clip_norm structures to ensure it runs without error (primarily) and productes correct results (secondarily)

File: jax_privacy/clipping.py [hunk 27]
```diff
@@ -131,7 +144,37 @@ def clip_pytree(
...
+        '_ClippedTreeAndNorm', ['clipped', 'norm']
+    )
+    results = jax.tree.map(
+        lambda c_norm, p_tree: _ClippedTreeAndNorm(
+            *recursive_clip(p_tree, c_norm)
+        ),
+        clip_norm,
+        pytree,
+    )
+    is_clipped_tree_and_norm_node = lambda x: isinstance(x, _ClippedTreeAndNorm)
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: Let's just call this "is_leaf", then the two tree.maps should each fit on one line below.

File: jax_privacy/clipping.py [hunk 28]
```diff
@@ -131,7 +144,37 @@ def clip_pytree(
...
   """
+  # -- Per-layer Norm --
+  if not jnp.isscalar(clip_norm):
+    recursive_clip = functools.partial(
+        clip_pytree,
+        rescale_to_unit_norm=rescale_to_unit_norm,
+        nan_safe=nan_safe,
+        return_zero=return_zero,
+    )
+    _ClippedTreeAndNorm = collections.namedtuple(
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I'd move this definition outside of the scope of this function.  Rather than map the raw results of recursive_clip, just have clip_pytree always return this named tuple

File: jax_privacy/clipping.py [hunk 29]
```diff
@@ -547,8 +596,16 @@ def clipped_fun_one_group(*args, is_padding_example, **kwargs):
...
     norm_bound = float(grid_scale)
+  elif rescale_to_unit_norm:
+    # If every leaf is rescaled to unit norm, the worst-case global norm is
+    # the square root of the total number of leaves in the PyTree.
+    num_leaves = len(jax.tree.leaves(l2_clip_norm))
+    norm_bound = math.sqrt(num_leaves) / normalize_by
   else:
-    norm_bound = (1.0 if rescale_to_unit_norm else l2_clip_norm) / normalize_by
+    norm_bound = (
+        sum(leaf**2 for leaf in jax.tree.flatten(l2_clip_norm)[0]) ** 0.5
```
Comment (ryan112358) [inline jax_privacy/clipping.py]: I guess jax.tree.leaves is slightly cleaner than jax.tree.flatten here

File: tests/clipping_test.py [hunk 30]
```diff
@@ -435,5 +543,85 @@ def fun(batched_data):
     self.assertEqual(aux_4d.shape, (15, 4, 9, 10))
 
 
+class ClippedFunPerLayerTest(parameterized.TestCase):
+
+  @parameterized.named_parameters(
```
Comment (ryan112358) [inline tests/clipping_test.py]: Same here

**Review by ryan112358 — CHANGES_REQUESTED** [review]

**Review by ryan112358 — APPROVED** [review]
